### ISSUE TITLE
Sidekiq or sidekiq-cron is bloating memory

### ISSUE THREAD

**Issue by Flaburgan** [diaspora/diaspora#6763]
I open an issue to be able to set this as a blocker for 0.6. This is a regression since we replaced sidetiq by sidekiq-cron and bumped sidekiq from 3 to 4, not sure which action caused it.

See the original discussion on the [pull request](https://github.com/diaspora/diaspora/pull/6616#issuecomment-182231026)

**Comment by denschub** [comment]
Cannot reproduce and apparently nobody seems to be responsible for providing a memory profile. Since you've opened this issue, it is your responsible to do so.

**Comment by Flaburgan** [comment]
Yep @cmrd-senya gave me this commit: https://github.com/cmrd-senya/diaspora/commit/c32809956224804930615df8055c25b83ded2bc9 I'll apply it and we will see.

**Comment by Flaburgan** [comment]
Btw, I'm running postgres, and here is my redis version: `Redis server v=3.0.7 sha=00000000:0 malloc=jemalloc-3.6.0 bits=64 build=49ffd7ce21c94625`

What about you @jaywink ?

**Comment by cmrd-senya** [comment]
I've managed to produce a [memory profile](https://we.riseup.net/assets/286085/heap200.json.xz).

To do so I ran two VMs provisioned with vagrant ([diaspora-replica](https://github.com/joebew42/diaspora-replica)) with some simple automated federation exchange - user discovery and sharing messages.

Then I followed the heap analisys way from [this](http://www.be9.io/2015/09/21/memory-leak/) wonderful article. I created 3 dumps of the heap after each 100 jobs processed. Then, using the article author's [tool](https://github.com/Shuttlerock/memory-dump-analyzer) I got the [object list](https://gist.github.com/cmrd-senya/72d5b89d1b1cf707e0cc) which is most probably represents leaked objects.

I am unsure what to do next. All the objects in the list point to different gems' code. Does it mean the leaks are there? Or it is just correct objects which live for longer for some reason?

I'll try to build a graph basing on the profile to track by references, which objects refer to the "leaked" objects to find their parent object which belong to the diaspora* application.

**Comment by cmrd-senya** [comment]
Some real word profile would be nice to have. If you can, @Flaburgan, please try this https://github.com/diaspora/diaspora/commit/db0a4e5f242ce092c8bb65ee5322868261b5cae6 on your pod.

**Comment by Flaburgan** [comment]
For how long should I let this commit run?

**Comment by cmrd-senya** [comment]
23.03.2016 19:12, Fla пишет:
>
> For how long should I let this commit run?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/diaspora/diaspora/issues/6763#issuecomment-200416807>
>

You have to get no less than three heap dumps. However, I think, it's
better if it takes a period where a significant memory usage growth can
be noticed. What time does your pod need to grow by a 1Gb from the
initial state?

**Comment by Flaburgan** [comment]
>What time does your pod need to grow by a 1Gb from the initial state?

I would say 20 seconds :p It raises up to 2GB in a few minutes. I'm going to add it now and ping you around 10:30pm on IRC this evening to see what's going on.

**Comment by Flaburgan** [comment]
>What time does your pod need to grow by a 1Gb from the initial state?

I would say 20 seconds :p It raises up to 2GB in a few minutes. I'm going to add it now and ping you around 10:30pm on IRC this evening to see what's going on.

**Comment by Flaburgan** [comment]
So, here are the data: https://diaspora-fr.org/heap100.json to 600 actually, others coming soon.

**Comment by cmrd-senya** [comment]
Here are the objects count diff for @Flaburgan's 200 and 600 heap dumps: https://gist.github.com/cmrd-senya/6ffde20c4a227d2cc61f

**Comment by cmrd-senya** [comment]
I summed up all the "memsize" fields of the heap600.json and got 147863516 which is 141Mb. It doesn't correspond to the reported growth rate. May be the leak is in the other process, which has it's own heap?

**Comment by Flaburgan** [comment]
So, surprisingly, the memory leak doesn't appear since I updated my server on the 23rd of March.
- The memory leak was here since https://github.com/diaspora/diaspora/pull/6616 was merged
- When https://github.com/diaspora/diaspora/pull/6586 was merged, I felt a bit nervous about such a big data remove so I wanted feedback before updating d-fr, so no more upgrade
- The 23rd of March, I rebased d-fr on the last develop commit applying all the changes.
- The 24th of March I restarted d-fr with the commit provided by @cmrd-senya to try to monitor the memory usage. Since that day, the leak looks not here anymore.

@jaywink on which commit your develop pod is running? Is the memory leak still there? If so, can you try to upgrade a few commits by a few commits, to try to understand what solved the problem?

**Comment by jaywink** [comment]
I've bumped the HQ pod to 0.5.99.0-p0cbb1233 (develop head now). Lets see if it stays stable this time.

**Comment by jaywink** [comment]
Well, it's not looking too positive for this pod atm, since bumping past sidekiq 4 commit again (I had rolled back to the one before that to stabilize it), this is what memory usage looks like on the server the last month. It is currently going way higher than the monthly maximum otherwise.

![](http://i.imgur.com/hj2DpJg.png)

The drops earlier are not restarts btw, not all of them at least. I think it is the second pod which occasionally decides to restart memory or restart sidekiq or something.

The HQ pod is now at 7.3G. If it goes over 10G I'll need to implement an automatic restart for it since I run many other apps on the server too.

**Comment by denschub** [comment]
Instead of randomly sharing the same numbers over and over again and complaining that sidekiq leaks memory, please provide actual memory statistics like mentioned multiple times in this issue and #6616. Your screenshots of memory usage statistics are not helpful at all, but some actual information may be.

@Flaburgan I assume you have checked that sidekiq is the process using a lot of memory and not something else, yes? Back in the old PR, I saw only one worker used a lot of memory. What was this worker doing? Were there export processes running?

**Comment by cmrd-senya** [comment]
@jaywink, if you keep seeing the memory leak, ping me so I assist you to profile the memory usage.

**Comment by jaywink** [comment]
@cmrd-senya Don't want to be unhelpful to anyone here so I already implemented an automatic restart once sidekiq reaches 7GB. Happens approx twice a day.

But if you have a commit to pull in that applies to latest develop and feel it might provide some information, sure, let me know.

It's probably just my server setup or something though.

**Comment by cmrd-senya** [comment]
You can try this one: https://github.com/diaspora/diaspora/commit/db0a4e5f242ce092c8bb65ee5322868261b5cae6

It'll create heap dumps after each 100 jobs processed. Then the heap dumps may be analyzed.

**Comment by Flaburgan** [comment]
Okay so I have something **really** weird. Here is the memory usage for the last 30 days on d-fr:
![capture du 2016-04-03 21-32-07](https://cloud.githubusercontent.com/assets/930064/14234462/9b39c258-f9e3-11e5-8ba8-61b9454afd66.png)

As said, on the 23rd of March, I rebased d-fr on develop. I then applied @cmrd-senya's commit, and restarted. As you can see, the memory leak was not there anymore. Everything was back to normal, so the 31th of March I removed senya's commit and restarted the pod. And boom, the leak is still fixed, but the memory usage grew back to a level slightly over what I had **before the bump to sidekiq 4**. With the commit, the memory usage was slightly under what I had before the bump. I will restart now without changing anything and see to which level the memory goes back.

**Comment by Flaburgan** [comment]
Okay so I made some tests and it is the [manual call](https://github.com/diaspora/diaspora/commit/db0a4e5f242ce092c8bb65ee5322868261b5cae6#diff-fd9e64e4a154ab77edea808fa4285141R44) of GC.start which make the memory stable in @cmrd-senya's commit. I modified his patch to run it every 1000 sidekiq jobs instead of 100 and the memory of d-fr is now stable at 10% for 3 hours now instead of 40 to 60% without the patch. The CPU usage is way bigger though, for < 5% to > 20%. I guess we need to try and find the number of jobs we want to wait before calling the GC.

**Comment by Flaburgan** [comment]
(Or to find why we need to manually call it...)

**Comment by Flaburgan** [comment]
So I tried every 1000, 10000, 10... jobs nothing change, 2 of my 8 CPU are at 100% with this sidekiq.rb. I also tried to remove the mutex and simply call the garbage collector 2% of the time with
`
def call(worker_instance, item, queue)
   yield
   if  rand(50) % 49 == 0
     GC.start
   end
 end
`
but there again 2 cores at 100%. So I reverted to the default sidekiq.rb, CPU are fine, and memory jumped back from 10% to 30% immediately. I don't know what to do now, if anybody wants to help...

**Comment by denschub** [comment]
I'm running an unpatched `develop` head on Geraspora vor 14 hours without any "leaks".

**Comment by cmrd-senya** [comment]
Is it possible there are some differences in the GC tuning? Are you
using different builds of Ruby? Maybe they are provided by different
vendors and therefore have a different set of patches?

**Comment by denschub** [comment]
Vanilla 2.3.0.

**Comment by cmrd-senya** [comment]
Earlier @Flaburgan said
> I use ruby-2.2.4 myself (from rbenv).

Ruby version difference may play some role I think.

**Comment by denschub** [comment]
It's funny how everyone here is completely ignoring me. I'll try again:

@Flaburgan What jobs is the "leaking" process doing? If all, start a dedicated worker per queue (but set the concurrency to 1 first...) and check what queue causes the bloat. Manually triggereing the GC is not an option since that adds totally unnecessary GC pause.

Can we please try to get over the whiny "oh my god it is so baad" and "oh look, I've found yet another workaround!" phase of this issue and try to actually **solve** it?

**Comment by jaywink** [comment]
@denschub you're the only one here who keeps constantly whining.. not everyone is as awesome as you so maybe chill a bit.

Also, @Flaburgan, I thought you didn't have a problem any more, or at least not memory going through the roof?

@cmrd-senya I applied the commit on top of current develop and noticed that memory usage was down to under 1 gig. Found out the reason - sidekiq keeps dying a few times an hour. Will remove the profiling commit to see if that was the cause. Otherwise the pod is fully vanilla diaspora with ruby 2.2.4 etc.

The heap dumps seem to contain lots of data content as well so sending you a link in private. Probably not very useful anyway since sidekiq never really gets up to speed. The sidekiq segfault paste is [here](http://pastebin.com/ygV8KR5m).

**Comment by jaywink** [comment]
Adding more unhelpful stuff, but not sure that it matters but the processed / failed ratio is pretty high on the HQ pod (top) compared to for example my own (below).. Probably because each "10 minutes online" pod creates a connection with the HQ account and the HQ account will forever try to deliver everything to it - though there shouldn't be that much stuff going out anyway.. The HQ sidekiq numbers are from when I moved it to it's own redis DB when sidekiq 4 got bumped. The ratio was probably as bad before.

![selection_252](https://cloud.githubusercontent.com/assets/1174866/14363880/ff12064a-fd0e-11e5-8973-ebd6bcfd33d4.png)
![selection_253](https://cloud.githubusercontent.com/assets/1174866/14363884/0219c8dc-fd0f-11e5-959d-03086d0d91aa.png)

**Comment by Flaburgan** [comment]
@denschub the leak is not there anymore, as said in https://github.com/diaspora/diaspora/issues/6763#issuecomment-201225763 

I was since then tempted to close this issue, but seeing the nice improvement on memory usage that senya's commit added, I kept it open to see if we could do something. I also wanted to wait for a confirmation from someone else that the issue is gone for them too.

**Comment by denschub** [comment]
Adding explicit GCing with senyas commit is not really a solution, and I'd really like to identify possible issues if there are any.

I expect the http-workers to take quite a lot of memory, but I wanted to confirm that, which is why I asked. I'm running sidekiq4 now for almost 24 hours without any issues, but my memory limits are beyond the range of what a sidekiq process in theory could even consume, which is why I need someones help here. Jason does not seem to be willing at all, so I'd love to have you set up a dedicated worker per queue to check if there is one worker-set consuming a lot more memory than all others do or something.

I still doubt we have a leak here (especially since GCing helps), but we might have a bloat issue somewhere. I'd like to at least narrow it down to a single worker so we have something to work with. Otherwise, this issue is a really useless thing because "there is *something* *somewhere* wrong" is never a good starting place...

Basically, this would mean running 12 sidekiq processes (`mkdir pids` first):

```
bin/bundle exec sidekiq -q http -p pids/http
bin/bundle exec sidekiq -q http_service -p pids/http_service
bin/bundle exec sidekiq -q socket_webfinger -p pids/socket_webfinger
bin/bundle exec sidekiq -q export -p pids/export
bin/bundle exec sidekiq -q maintenance -p pids/maintenance
bin/bundle exec sidekiq -q delete_account -p pids/delete_account
bin/bundle exec sidekiq -q receive -p pids/receive
bin/bundle exec sidekiq -q receive_salmon -p pids/receive_salmon
bin/bundle exec sidekiq -q receive_local -p pids/receive_local
bin/bundle exec sidekiq -q dispatch -p pids/dispatch
bin/bundle exec sidekiq -q photos -p pids/photos
bin/bundle exec sidekiq -q mail -p pids/mail
```

If you see a sidekiq process consuming a lot of money, note the pid and check the files in `pids/` to see which queue caused the bloat. That would be at least a bit helpful.

**Comment by denschub** [comment]
What I am aiming for: even if you don't have excessive memory usage anymore, sidekiq4 shouldn't really use more memory than sidekiq3. In fact, it should consume a lot less memory, so something may be wrong.

**Comment by cmrd-senya** [comment]
I think I can simulate intensive traffic exchange between a couple of
test pods, if this is relevant to what you want to test.

**Comment by Flaburgan** [comment]
So I saw with Dennis, I disabled sidekiq inside diaspora* by commenting https://github.com/diaspora/diaspora/blob/develop/config/eye.rb#L26-L39 and then launched the processes as asked above.

After 3 hours, here is the result:

process | residential memory
-----------|-------------------------
http | 1546M
delete_account | 346M
unicorn0 | 293M
http_service | 292M
unicorn2 | 290M
unicorn3 | 289M
mail | 285M
receive | 282M
unicorn1 | 281M
receive_salmon | 253M
receive_local | 234M
dispatch | 232M
photos | 220M
maintenance | 220M

etc

**Comment by Flaburgan** [comment]
```
curl -V
curl 7.32.0 (x86_64-pc-linux-gnu) libcurl/7.32.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 libssh2/1.2.8 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smtp smtps telnet tftp 
Features: AsynchDNS GSS-Negotiate IDN Largefile NTLM NTLM_WB SSL libz TLS-SRP
```

**Comment by Flaburgan** [comment]
22 hours later, here is the status:

pid | process | residential memory
-----|-----------|-------------------------
30535 | http | 2223M
30732 | http_service | 490M
 | vines | 385M
31936 | maintenance | 352M
32233 | delete_account | 348M
32422 | receive  | 335M
 | unicorn0 | 296M
1180 | mail | 292M
 | unicorn3 | 289M
 | unicorn2 | 286M
32621 | receive_salmon | 286M
 | unicorn1 | 282M
552 | receive_local | 244M
748 | dispatch | 242M
951 | photos | 225M
30945 | socket_webfinger  | 220M
31254 | export | 220M

**Comment by cmrd-senya** [comment]
Do we need to profile http worker with ObjectSpace now, @denschub?

**Comment by denschub** [comment]
Thanks @Flaburgan. That's actually quite useful.

> Do we need to profile http worker with ObjectSpace now

Nah. This has to be bloat by Hydra (which is the reason @jhass created #5217) or a leak by curl. In any case, analyzing the worker has no chance of a useful outcome.

Well, the second I wrote this, my mind reminded me of something... One possible optimization could be not loading the people inside the worker, but rather the actual hydra processor. Sadly, I am not sure if this would actually work, since I know that typhoeus is doing concurrent requests, but I'm not sure how. In theory, a bad suboptimal concurrency implementation could leak those objects, but measuring that is not that easy. I'll have to dig into that to check if we could pass IDs instead of the people object into the processor since that could reduce the memory footprint by quite some bytes, especially since our default `typhoeus_concurrency` is set to 20.

**Comment by Flaburgan** [comment]
24 hours after my previous message, 48h since the beginning, here is the status:

pid | process | residential memory
-----|-----------|-------------------------
30535 | http | 2643M
30732 | http_service | 585M
31936 | maintenance | 483M
32422 | receive  | 455M
 | vines |423M
32233 | delete_account | 349M
32621 | receive_salmon | 323M
1180 | mail | 305M
 | unicorn0 | 291M
 | unicorn3 | 288M
 | unicorn2 | 285M
 | unicorn1 | 281M
552 | receive_local | 264M
748 | dispatch | 248M
951 | photos | 227M
31254 | export | 221M
30945 | socket_webfinger  | 220M

Please tell me if there is anything else I can run to help you debug.

**Comment by denschub** [comment]
No, you can go back to your old setup now. I'll think about what I've written before. Thank you very much again!

**Comment by Flaburgan** [comment]
Just fyi because I didn't restart the pod at the moment, `http` is now using 3738M of res memory, `maintenance` is using 1043M and `http_receive` 1034M. Vines is fourth with 679M. I'm wondering if there is something we can do to monitor the `maintenance` process.

**Comment by denschub** [comment]
#6830 contains a Sidekiq update with an interesting changelog: https://github.com/mperham/sidekiq/blob/5e15c06d3913e645ad966e8dc91f08e98ab5335d/Changes.md#412

Not sure what that means, though, so far, I did not have enough time to look into that.

**Comment by cmrd-senya** [comment]
I can try to build the test env where it is getting reproduced. I'm not sure if it worth the efforts though. Tell me, if it could help for that issue solution.

**Comment by Flaburgan** [comment]
So I may have something new about this. Here is the interesting usage of memory by d-fr in the last 30 days.

![sidekiq](https://cloud.githubusercontent.com/assets/930064/15363181/28298728-1d18-11e6-82f5-ae2e1a55749d.png)

At the beginning, d-fr was running following @denschub indications (see https://github.com/diaspora/diaspora/issues/6763#issuecomment-208409841) that means no launch of sidekiq in script/server and manual launch of each 12 processes separatly. As you can see, even if the memory usage constantly rise, it was a really slow rise. I went back to the normal way of launching diaspora* on the 13th of May, and here we can see the memory exploding again. I waited for a few days and then updated d-fr to the last develop code, which means with redis and sidekiq bump. As you can see, this doesn't seem to change anything.

So it looks to me that the leak can be mainly solved by a correct configuration. So @denschub can you share all your sidekiq settings? (Nb workers, concurrency etc) I'll try to adapt mine to yours and see if the problem is still there.

Mine are:
sidekiq:
concurrency: 5
retry: 12
namespace: "diaspora-fr"
backtrace: 20

unicorn_worker: 4
unicorn_timeout: 90
sidekiq_workers: 2

typhoeus_concurrency: 50

@jaywink would be nice to know yours too.

**Comment by Flaburgan** [comment]
I'm also going back to the default diaspora* settings, to see how memory goes.

**Comment by jaywink** [comment]
https://pod.diaspora.software is now at latest devel commit.

These are the only explicitly set settings re sidekiq and redis:

    sidekiq:
        concurrency: 5
        backtrace: false
    redis: redis://127.0.0.1:6379/1

**Comment by Flaburgan** [comment]
So, the good news is, back to the default diaspora* settings, I have a similar memory usage to the one I had with sidekiq launched separately (constant slow rise):
![memory](https://cloud.githubusercontent.com/assets/930064/15572256/c1e9d93a-233f-11e6-9dd0-e6934bee17b5.png)

**50% still means 12GB of memory used by sidekiq**, for a pod with [less than 200 users active per month](https://diaspora-fr.org/statistics), that's way too much.

Especially, when you look at the memory usage for the last 12 months, we clearly see that before the bump to Sidekiq 4 and replacement of sidetiq by sidekiq-cron, the memory usage was lower.
![memory2](https://cloud.githubusercontent.com/assets/930064/15572258/c20b786a-233f-11e6-899f-59ec7b8834d7.png)

Maybe we can try something else than sidekiq-cron, which is not that active btw. [Sidekiq-scheduler](https://github.com/Moove-it/sidekiq-scheduler) for example?

**Comment by Flaburgan** [comment]
[Sidetiq](https://github.com/tobiassvn/sidetiq) also got an update, and is now requiring [celluloid 0.17 and sidekiq 4](https://gemnasium.com/tobiassvn/sidetiq) so maybe we can also try to switch back and see how memory is going.

**Comment by jhass** [comment]
https://github.com/tobiassvn/sidetiq/pull/154 is hanging around for two months again already.

**Comment by Flaburgan** [comment]
@jhass yes I was proposing to sidetiq back bot for the release but just to see if it fixes the regression (which would mean sidekiq-cron is the problem) or not (sidekiq 4 is the problem).

**Comment by jaywink** [comment]
No change in memory usage pattern at pod.diaspora.software, since bumping to latest develop. Sidekiq still tops 10G approx once per day, where I've set it to be killed.

![selection_265](https://cloud.githubusercontent.com/assets/1174866/15586926/878d96f0-238f-11e6-99ef-593af3dca399.png)

**Comment by Flaburgan** [comment]
Good news, with the new federation, the memory is stabilizing at ~50% (12GB) when it was >60% with the previous develop code. It was ~27% before the regression though so the problem is not solved. I'm going to restart d-fr with ruby 2.3.1 and see if it changes anything.

**Comment by denschub** [comment]
Now that #6918 is merged, I'm curious to know you experience less memory usage!

**Comment by Flaburgan** [comment]
I'm currently monitoring it really carefully, it's too soon to tell but it looks very promising. It would be cool to have feedback from people still running ruby 2.2 too (ping @jaywink ? )

**Comment by Flaburgan** [comment]
So here is the result of my recent tests:
![memory-dfr](https://cloud.githubusercontent.com/assets/930064/16944005/e323ad90-4d9e-11e6-9c86-4ded7a94d963.png)

I'm using the default settings of diaspora*, concurrency of typhoeus is 20.
1. Old federation, ruby 2.2, default settings
2. New federation, ruby 2.2. The memory is stabilizing lower at 50% (12GB), it's already better
3. New federation, ruby 2.3
4. New federation with https://github.com/diaspora/diaspora_federation/pull/37

So the curl patch does improve nicely the memory usage! I'll wait for a few more days and make another report.

**Comment by denschub** [comment]
Uh, I did not expect to get a 50% memory usage decrease out of my patch. 😮

**Comment by jhass** [comment]
We can also document some GC tuning apparently, I'm running Ruby 2.3 with the latest develop and the following environment variables set for a couple of days:
```
RUBY_GC_HEAP_GROWTH_FACTOR=1.1
RUBY_GC_HEAP_OLDOBJECT_LIMIT_FACTOR=1.2
RUBY_GC_MALLOC_LIMIT_GROWTH_FACTOR=1.1
RUBY_GC_OLDMALLOC_LIMIT_GROWTH_FACTOR=1.1
```

**Do not blindly set these**. Read on.

While I get a lower and more stable memory usage with these, the values I picked there are very very aggressive. On my defacto single user pod I do not notice any real slow downs, but I would expect on bigger pods it might get noticeable, that is slower response times, slower sidekiq processing that might eventually lead to backlog. If you want to experiment  with these, do take your time to understand what they do first. A good write up is at http://thorstenball.com/blog/2014/03/12/watching-understanding-ruby-2.1-garbage-collector/ which you should read fully, http://collectiveidea.com/blog/archives/2015/02/19/optimizing-rails-for-memory-usage-part-2-tuning-the-gc/ complements it with some experiences and considerations.

But if enough people experiment with this, we eventually we might gather enough experience in the community to be able to recommend settings for small, medium and large setups.

**Comment by jaywink** [comment]
![selection_285](https://cloud.githubusercontent.com/assets/1174866/16995368/a83bb814-4eb4-11e6-8db5-f3aa09e51352.png)

The HQ pod started not restarting itself (as I've told it to) every time Sidekiq tops 10G after the last code deploy, which was ~1st of July. Well, it did top 10G once after that, not sure why - maybe the initial restart was unclean or something.. Anyway, now it seems more stable.

Concurrency is 5 with no backtrace - default settings otherwise. iliketoast.net runs on similar 5 concurrency, with the code at the commit just before Sidekiq 4.

Current memory usage is as follows and very stable:
![selection_286](https://cloud.githubusercontent.com/assets/1174866/16995502/661ec06a-4eb5-11e6-832c-13bcbff39f70.png)

So on my server, memory usage on concurrency of 5 is higher still than Sidekiq 3, but at least it's not trying to kill the server since last deploy. Though it is slowly rising, which iliketoast.net sidekiq doesn't do.

I'll update HQ to the current develop head next.

**Comment by Flaburgan** [comment]
@jaywink can you also tell us which Ruby you are using?

**Comment by jaywink** [comment]
Whatever is in `.ruby-version` in the code repo is always used for both. Looks like HQ has ended on 2.2.4 and iliketoast.net on 2.2.1, probably because I haven't deployed it in a while.

**Comment by jaywink** [comment]
Well, seems stable enough - and lower memory usage than iliketoast.net.

Currently, and quite stable over last ~8 days:

![selection_289](https://cloud.githubusercontent.com/assets/1174866/17221021/99187984-54fa-11e6-9383-080774d0f0d9.png)

**Comment by denschub** [comment]
Cool.

**Comment by Flaburgan** [comment]
Nice so that means the improvement is there even with 2.2 :)

d-fr is still slowly increasing but only uses 35% of memory after 15 days.

**Comment by Flaburgan** [comment]
The memory usage of d-fr now looks quite stable for one week around 35%. This is still higher than the usage with 0.5, but it looks like we have no real leak anymore. Maybe we can keep this issue open because there is still room for improvements and tweaking as @jhass said, but it would feel okay for me if we remove the 0.6 milestone on that issue. Thanks a lot @denschub for the patch!

**Comment by koehn** [comment]
I'm experiencing the memory bloat as well as CPU. My pod is using 1.6 GB of RAM, and is consuming one core nearly all the time on my machine. It's a real problem. I'm happy to give diagnostics as needed; Jason Robinson directed me here.

**Comment by Flaburgan** [comment]
@koehn please tell us what are your settings? (typhoeus concurrency, sidekiq concurrency, nb of sidekiq workers, retry...)

**Comment by koehn** [comment]
typhoeus_concurrency: 20

sidekiq concurrency: 25

sidekiq retry: 10

sidekiq backtrace: 15 

unicorn_rorker: 2

embed_sidekiq_worker: true

sidekiq_workers: 1

I used to receive all tags, but reduced the tag count to about in hopes that would reduce CPU load, but changing the setting and restarting the pod had no effect.

**Comment by Flaburgan** [comment]
Default sidekiq concurrency [is 5](https://github.com/diaspora/diaspora/blob/develop/config/defaults.yml#L16), you can but that back and your memory usage should go down.

**Comment by koehn** [comment]
From `diaspora.yml`(emphasis added):
```
      ## Number of parallel threads Sidekiq uses
      ## If you touch this please set the pool setting
      ## in your database.yml to a value that's at minimum
      ## close to this! The default value is 5 but you can safely
      ## increase it to 25 and more on a medium sized pod.
      ## This applies per started Sidekiq worker, so if you set it to
      ## 25 and start two workers you'll process up to 50 jobs in parallel.
```

So if I set it back to 5 is my pod going to suffer the same performance problems that caused me to increase it in the first place (at your recommendation, IIRC)?

**Comment by Flaburgan** [comment]
It's all about performance VS memory usage so you have to find the best setting for your pod, based on the active users and the memory your server has.

**Comment by jaywink** [comment]
@koehn I'm not an expert but I'm pretty sure this should be `false` ?

    embed_sidekiq_worker: true

**Comment by koehn** [comment]
Dunno; it was the default value so I left it alone.

**Comment by SuperTux88** [comment]
the default is false, and as far as I remember it was always false. But it is true by default in development mode, but it is recommended to set it to false (default) in production.

**Comment by koehn** [comment]
It used to be true (there are no secrets from Git): 

https://github.com/diaspora/diaspora/commit/79a79d65d684b6cfa93e1d2dde9ac7abcfb7a481

**Comment by jaywink** [comment]
[Actually no it was false](https://github.com/diaspora/diaspora/commit/79a79d65d684b6cfa93e1d2dde9ac7abcfb7a481#diff-2c96c573eaf0842befdbce1bb3756645R42) - it's just that the diaspora.yml.example used to be more confusing ;)

**Comment by koehn** [comment]
In any case, my `top` sorted by memory looks like this these days (after making the changes mentioned above):

```
top - 16:27:27 up 98 days,  7:53,  2 users,  load average: 0.56, 0.60, 0.66
Tasks: 337 total,   1 running, 336 sleeping,   0 stopped,   0 zombie
%Cpu(s):  6.1 us,  1.3 sy,  0.0 ni, 91.0 id,  0.4 wa,  0.0 hi,  1.2 si,  0.0 st
KiB Mem:  16176184 total, 15886364 used,   289820 free,   204260 buffers
KiB Swap:    94204 total,    94204 used,        0 free.  5957856 cached Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                
22487 diaspora  20   0 3599064 2.424g   9392 S   0.0 15.7 197:23.62 ruby                                                                                                                                   
```

That's a whole lot of RAM leaking to run a pod.

**Comment by koehn** [comment]
Here's the `smem` output:

```
  PID User     Command                         Swap      USS      PSS      RSS 
22487 diaspora sidekiq 4.1.4 diaspora [1 o        0  2536844  2538113  2545332 
```

**Comment by denschub** [comment]
Let me summarize what I wrote on 27th of March and in some more recent comments:

* Stop posting memory usage numbers unless you are explicitly asked to do so. @Flaburgan 's numbers per queue for example were requested and very useful, they actually led me toward a solution for parts of a problem we had. Pasting `top`/`smem` outputs is only creating noise. Without detailed information about the memory contents and the full environment, these numbers are totally useless and make this issue look like everyone is whining.
* No, there is no **leak**. A memory leak is indicated by a constantly growing memory usage without any limit. We've had one in the past by reusing cURL sockets (technically that was not even a leak either since it eventually could have reached its limit, but the numbers are way to high on this), which got fixed. Even for @jaywink, memory consumption seems to be stable now. Nobody in the active development team was able to find an leak, so unless you can prove that there is a leak, there are none.
* Yes, Sidekiq eats quite a lot of memory. However, compared to, for example GitLab, our memory consumption is quote low if you match up the concurrency settings.
* If you want to reduce your memory footprint, reduce Sidekiqs concurrency and reduce the size of the database connection pool.

If you actually want to contribute to this issue in any way:

* Figure out if we even need more than, let's say, 2 sidekiq threads. Our new federation tasks are really fast compared to the old ones and we might not even run into queuing problems with smaller numbers. 5/25 were picked ages ago, with basically everything inside diaspora\* looking different. One could, for example, draw charts with average job wait times. If workers are idle most of the time, we could try decreasing the concurrency to increase the worker worktime while keeping the peak job time below 5 minutes, for example.
* If you feel adventurous, see https://github.com/diaspora/diaspora/issues/6763#issuecomment-233593973 from July. There's an explicit warning and *"But if enough people experiment with this, we eventually we might gather enough experience in the community to be able to recommend settings for small, medium and large setups."*, which is still true.
* Understand how memory profiling in Ruby cleanly works and dig further into the memory traces. We know that a lot of memory is allocated inside ActiveRecord bindings, but this information is totally worthless without knowing what's inside those bindings. Also, figure out on how we can make this more efficient. Don't ask here, we don't have an idea. However, please write a blog post or something about this since I would be really interested in learning that.

If you feel like everything sucks, feel free to:

* Submit Pull Requests to Sidekiq to help them improve their memory footprint.
* Investigate alternative Queuing solutions. Compare and benchmark them. Port diaspora\*.
* Rewrite our workers using, for example, goworker. Also, figure out how to build and deploy them for all podmins without having them setting up an entire Go build stack.
* Build our very own and tailored solution for background processing, with performance and a low footprint in mind. Also, figure out how to build and deploy them for all podmins without having them setting up an entire Rust/Crystal/Go/WhateverYouChose build stack.

In any case, **stop using this issue for "I am having memory bloat as well" posts**. I will delete all of them that ever appear below this comment. Our GitHub issue tracker is not a support platform, and I don't want to discuss configuration issues here at all. If we figure out our default settings are bad (which very well may be), we should change them. If we figure out we can give podmins a better idea of what they should set their concurrency to, we should add that stuff to the wiki. The only things that belong into this issue, if anything, are discussions regarding memory improvements we could perform on our side.

**Comment by koehn** [comment]
Right. Sorry if my comments inconvenienced you; I was directed to post here by several developers on the team. If they're unwelcome, then fine. I understand how frustrating developing FOSS projects can be, and have no intention of making a difficult situation worse. I don't speak Ruby, so I'm unlikely to be offering a fix to this problem; indeed I'm working on an alternative platform that is completely serverless.

**Comment by koehn** [comment]
FWIW I've been running for the past week or so on 2 sidekiq workers without problems.

**Comment by 001101** [comment]
Why not just embedding libjemalloc as it GitLab or Sandstorm does?

locate libjemalloc.so
/opt/gitlab/embedded/lib/libjemalloc.so
/opt/gitlab/embedded/lib/libjemalloc.so.2
/var/sandstorm/apps/35f82e4bf3ef0683558336494a3dc9d6/usr/lib/x86_64-linux-gnu/libjemalloc.so.1

**Comment by SuperTux88** [comment]
I didn't know jemalloc, but I tried this now for a week on the HQ pod since I needed to restart sidekiq there almost daily after the G+ rush, and it looks like it works pretty stable. Sidekiq was arround 500-700MB while it was multiple GB before:

![ram usage](https://user-images.githubusercontent.com/458548/47827976-1d7c3880-dd81-11e8-9a21-99510ad1cfd4.png)

I think we can just check in `script/server` if it is available and `LD_PRELOAD` it. And we then can add it to the installation guides and next major upgrade guide, so podmins can install it :) I don't know if we can close this then, because this is just how sidekiq is, or if we still have a problem somewhere?

**Comment by denschub** [comment]
I did some testing as well, and while my sidekiq eventually always got stable, it's now stable at a lower point:

<img width="496" alt="screenshot 2018-11-01 at 02 59 11" src="https://user-images.githubusercontent.com/344777/47828183-333e2d80-dd82-11e8-8a0d-ab388b5f2461.png">

It seems like it's not a bad idea.

> I think we can just check in script/server if it is available and LD_PRELOAD it. And we then can add it to the installation guides and next major upgrade guide, so podmins can install it :)

Yeah, sounds good, why not. 👍 

> I don't know if we can close this then, because this is just how sidekiq is, or if we still have a problem somewhere?

I am all for closing issues. There is a "reopen" button if we need to use it.

**Comment by goobertron** [comment]
> I think we can just check in script/server if it is available and LD_PRELOAD it. And we then can add it to the installation guides and next major upgrade guide, so podmins can install it :)

Would it be a good idea also to do an HQ post offering this solution to any podmins who are struggling with Sidekiq loads?

**Comment by 001101** [comment]
Would also recommend to restart sidekiq each 4 hours or something like that and writing it to the docs, so the bloat will keep very low, it is just a simple cronjob or is it not recommend for some reason? Could not feel any difference.

`cd /home/diaspora/diaspora && bundle exec sidekiqctl stop /home/diaspora/diaspora/tmp/pids/sidekiq1.pid 30 && bundle exec sidekiq -d -L /home/diaspora/diaspora/log/sidekiq.log -C /home/diaspora/diaspora/config/sidekiq.yml -e production
`

**Comment by SuperTux88** [comment]
> Would it be a good idea also to do an HQ post offering this solution to any podmins who are struggling with Sidekiq loads?

I don't think we need an extra post for that. Adding it to the changelog and mention it in the "notable changes" section in the next release posts (also on the HQ account) should be enough. Podmins using the `script/server` probably don't need to do anything, maybe install the package, but on some systems it's already installed), and podmins with special solution would need a special solution for that too (which should be really easy for them if they know how their special solution works to just add it there, but would be much harder to describe in a post).

> Would also recommend to restart sidekiq each 4 hours or something like that and writing it to the docs, so the bloat will keep very low, it is just a simple cronjob or is it not recommend for some reason?

I wouldn't recommend that, and there is no reason to restart sidekiq regularly (especially when it runs with a much more stable memory usage with libjemalloc). I only restarted it when the memory usage was too high (which usually was once every few days), or when I update.

**Comment by goobertron** [comment]
> I don't think we need an extra post for that. Adding it to the changelog and mention it in the "notable changes" section in the next release posts (also on the HQ account) should be enough.

I was thinking of posting now, so that podmins know that help is available before the next major release.

**Comment by SuperTux88** [comment]
I added my PR to the next minor milestone, because there is no need to wait for a major release (it doesn't break without manually steps, the extra notes in the next major upgrade guide is only for podmins who missed it until then), and I also think that podmins with memory problems already have a solution that works for them at the moment and/or are subscribed to this issue.

**Comment by goobertron** [comment]
OK, thanks.
